haystack/components/preprocessors/markdown_header_splitter.py

```diff
@@ -0,0 +1,324 @@
+        logger.debug("Splitting text by markdown headers")
+
+        # find headers
+        matches = list(re.finditer(self._header_pattern, text))
+
+        # return unsplit if no headers found
+        if not matches:
+            logger.info("No headers found in document; returning full document as single chunk.")
+            return [{"content": text, "meta": {"header": None, "parent_headers": []}}]
+
+        # process headers and build chunks
+        chunks: list[dict] = []
+        header_stack: list[Optional[str]] = [None] * 6
+        active_parents: list[str] = []
+        has_content = False  # Flag to track if any header has content
+
+        for i, match in enumerate(matches):
+            # extract header info
+            header_prefix = match.group(1)
+            header_text = match.group(2).strip()
+            level = len(header_prefix)
+
+            # get content
+            start = match.end()
+            end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
+            content = text[start:end].strip()
+
+            # update header stack to track nesting
+            header_stack[level - 1] = header_text
+            for j in range(level, 6):
+                header_stack[j] = None
+
+            # skip splits w/o content
+            if not content:
+                # Add as parent for subsequent headers
+                active_parents = [h for h in header_stack[: level - 1] if h is not None]
+                active_parents.append(header_text)
+                continue
+
+            has_content = True  # At least one header has content
+            # get parent headers
+            parent_headers = list(active_parents)
+
+            logger.debug(
+                "Creating chunk for header '{header_text}' at level {level}", header_text=header_text, level=level
+            )
+
+            chunks.append(
+                {
+                    "content": f"{header_prefix} {header_text}\n{content}",
+                    "meta": {"header": header_text, "parent_headers": parent_headers},
+                }
+            )
+
+            # reset active parents
+            active_parents = [h for h in header_stack[: level - 1] if h is not None]
+
+        # return doc unchunked if no headers have content
+        if not has_content:
+            logger.info("Document contains only headers with no content; returning original document.")
+            return [{"content": text, "meta": {}}]
+
+        logger.info("Split into {num_chunks} chunks by markdown headers.", num_chunks=len(chunks))
+        return chunks
+
+    def _apply_secondary_splitting(self, documents: list[Document]) -> list[Document]:
+        """
+        Apply secondary splitting while preserving header metadata and structure.
+
+        Ensures page counting is maintained across splits.
+        """
+        if not self.secondary_split:
+            return documents
```

Comment:
this doesn't seem necessary since we already perform this check in the run method
```python
final_docs = self._apply_secondary_splitting(header_split_docs) if self.secondary_split else header_split_docs
```
so I think it's safe to remove. 

If it's only kept for mypy reasons please leave a comment explaining.